graphql_compiler/interpreter/typedefs.py

```diff
@@ -338,12 +338,140 @@ def project_neighbors(
+                    current_type_name, edge_info, current_token
+                )
+            )
+            # END OF BUGGY CODE
+        The above code is the victim of an unfortunate and difficult-to-find bug similar to
+        a race condition in multithreaded code:
+        - The current_token value is set to a new value in each iteration of the function's
+          outer "for data_context in data_contexts" loop.
+        - The generator in the buggy snippet refers to current_token -- but instead of referring
+          to "whatever is the value of current_token at the time the generator is defined,"
+          it instead refers to "whatever value current_token points to at each generator iteration".
+        - Since the outer loop's iterations and the buggy generator's iterations may occur in
+          an arbitrarily-iterleaved order, the buggy generator's current_token value changes
+          unpredictably during its execution therefore causing the bug.
+        - Since Python generators are not started until their first result is requested, the above
+          buggy snippet can't even guarantee that the first use of current_token has the intended
+          value -- the outer loop might have iterated before the generator was first invoked.
+
+        This bug is avoided (as in the working example code) by calling a function and supplying
+        the value of current_token to it. This creates a new Python scope (a so-called "closure")
+        that remembers the value current_token had when the closure was created. A succinct,
+        internet-search-friendly way to describe the cause of this bug is to say that in Python,
+        generators are not closures.
+
+        ## Hints supplied to this function refer to neighboring vertices
+
+        Hint kwargs such as used_property_hints, filter_hints, and neighbor_hints in this function
+        describe the desired structure of the *neighboring* vertices that this function produces,
+        instead of to the vertices supplied via the data_contexts input. For example, consider
+        the following query:
+            {
+                Foo {
+                    out_Foo_Bar {
+                        name @output(out_name: "name")
+                    }
+                }
+            }
+        The project_neighbors() call corresponding to the out_Foo_Bar edge traversal would contain
+        used_property_hints=frozenset({"name"}). This is because project_neighbors() here would
+        be called DataContexts whose current_token would point to Foo vertices, and the neighboring
+        vertices (ones along the outbound Foo_Bar edge) have their "name" property queried.
+
+        Args:
+            data_contexts: iterable of DataContext objects which specify the DataTokens whose
+                           neighboring DataTokens need to be loaded.
+            current_type_name: name of the vertex type whose neighbors need to be loaded. Guaranteed
```

Comment:
FWIW, I prefer the original language because of the explicit relationship to neighbors.  Maybe we could add something about the relationship to DataTokens to the end of the following sentence?

Reply:
Kept the original language and staged for commit with this parenthetical:

> (current_type_name names the concrete type of DataToken contained in the DataContext containers yielded by the data_context iterator.)